### PR TITLE
Alerting: Add relativeTimeRange from dataSource when using Resample expresions

### LINKED ISSUE #4144: [Prometheus data source] Set default format for "all" and "multi" to "pipe"
Prometheus only supports pipe format for templates so at the least the default should be changed; optionally, all other options can simply be discarded.

### PR DESCRIPTION
**Pull request by soniaAguilarPeiron**
**What this PR does / why we need it**:
This PR adds `relativetimeRange` in the `eval` request when using a `Resample` expresion.

### CODE REVIEW

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts [hunk 1]
```diff
@@ -78,10 +78,13 @@ export const queriesAndExpressionsReducer = createReducer(initialState, (builder
     })
     .addCase(updateExpression, (state, { payload }) => {
       state.queries = state.queries.map((query) => {
+        const dataSource = state.queries.find((alertQuery) => alertQuery.refId === payload.expression);
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:81]: Could we add a reducer's test for this case?

Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:81]: yes!

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts [hunk 2]
```diff
@@ -78,10 +78,13 @@ export const queriesAndExpressionsReducer = createReducer(initialState, (builder
...
+        const relativeTimeRange = dataSource ? dataSource.relativeTimeRange : getDefaultRelativeTimeRange();
         return query.refId === payload.refId
           ? {
               ...query,
               model: payload,
+              relativeTimeRange: relativeTimeRange,
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:87]: Do we need that for all types of expressions or just for `resample`?

**Review by peterholmberg — COMMENTED** [review]
Looks good to me, but I wonder if this is not something the backend can easily do instead? Maybe we can clear that before we merge this?

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx [hunk 3]
```diff
@@ -117,8 +118,9 @@ export const QueryAndExpressionsStep: FC<Props> = ({ editingExistingRule }) => {
 
   const onChangeQueries = useCallback(
     (updatedQueries: AlertQuery[]) => {
+      console.log('hola');
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:121]: 😆

**Comment by soniaAguilarPeiron** [comment]
> s good to me, but I wonder if this is not something the backend can easily do instead? Maybe we can clear that before we merge this?

@grobinson-grafana agreed to merge this fix to main.

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts [hunk 4]
```diff
@@ -78,13 +79,38 @@ export const queriesAndExpressionsReducer = createReducer(initialState, (builder
     })
     .addCase(updateExpression, (state, { payload }) => {
       state.queries = state.queries.map((query) => {
-        return query.refId === payload.refId
-          ? {
-              ...query,
-              model: payload,
-            }
-          : query;
+        const dataSource = state.queries.find((alertQuery) => alertQuery.refId === payload.expression);
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:82]: I would rename `dataSource` to `query`. 
I think there is one more edge case. What if the resample input is not a query but a different expression?

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts [hunk 5]
```diff
@@ -78,13 +79,38 @@ export const queriesAndExpressionsReducer = createReducer(initialState, (builder
...
+              };
+        } else {
+          return query;
+        }
+      });
+    })
+    .addCase(updateExpressionTimeRange, (state) => {
+      const newState = state.queries.map((query) => {
+        if (query.datasourceUid === ExpressionDatasourceUID) {
+          const dataSource = state.queries.find((alertQuery) => alertQuery.refId === query.model.expression);
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:103]: I think `dataSource` can be either query or expression

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts (hunk 4 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:82]: 💯, I will take a look on this edge case @konrad147

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts [hunk 6]
```diff
@@ -18,6 +19,34 @@ export interface QueriesAndExpressionsState {
   queries: AlertQuery[];
 }
 
+const findDataSourceFromExpressionRecursive = (
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:22]: 👏

Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:22]: Can we move this to a util file? I feel like it doesn't belong in the reducer and could be useful in other parts of alerting.

**Review by gillesdemey — APPROVED** [review]
LGTM, nice work!

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts (hunk 6 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:22]: Good idea, I will move it!

**Pull request merged**